android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt

```diff
@@ -0,0 +1,117 @@
+     * @param isAuthenticated whether the key requires user authentication to encrypt/decrypt
+     * @return an AES [SecretKey]
+     */
+    @Throws(
+        KeyStoreException::class,
+        IOException::class,
+        NoSuchAlgorithmException::class,
+        CertificateException::class,
+        UnrecoverableEntryException::class
+    )
+    fun getAESWithGCMSecretKey(keystore: String, alias: String, isAuthenticated: Boolean = false): SecretKey {
+        val spec = if (isAuthenticated) getGCMWithUserAuthenticationKeySpec(alias) else getGCMKeySpec(alias)
+        return getAESSecretKey(keystore, spec)
+    }
+
+    /**
+     * Get AES secret key for the given keystore, alias and spec. This method will return the existing key, or
+     * create a new key.
+     *
+     * @param keystore the keystore to get the secret key from
+     * @param spec the [KeyGenParameterSpec] spec to use to create the key if creation is needed.
+     * @return an AES [SecretKey]
+     */
+    @Throws(
+        KeyStoreException::class,
+        IOException::class,
+        NoSuchAlgorithmException::class,
+        CertificateException::class,
+        UnrecoverableEntryException::class
+    )
+    fun getAESSecretKey(keystore: String, spec: KeyGenParameterSpec): SecretKey = CipherLock.withLock {
+        // Attempt to fetch existing stored secret key from Android KeyStore
+        val keyStore = KeyStore.getInstance(keystore)
+        keyStore.load(null)
+        val entry = keyStore.getEntry(spec.keystoreAlias, null) as? KeyStore.SecretKeyEntry
+        val secretKey = entry?.secretKey
+
+        if (secretKey != null) return secretKey
+
+        // At this point, no secret key is stored so generate a new one.
+        val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, keystore)
+        keyGenerator.init(spec)
+
+        return keyGenerator.generateKey()
+    }
+
+    /**
+     * Check if the given keystore contains the alias
+     *
+     * @param keystore the keystore to check
+     * @param alias the alias of the key to check if it exists
+     * @return boolean indicating whether the alias exists in the given key store
+     */
+    fun contains(keystore: String, alias: String): Boolean = CipherLock.withLock {
+        val keyStore = KeyStore.getInstance(keystore)
+        keyStore.load(null)
+        return keyStore.containsAlias(alias)
+    }
+
+    /**
+     * Delete the given key alias from the keystore if it exists
+     *
+     * @param keystore the keystore to delete the alias from
+     * @param alias the alias of the key to delete
+     */
+    fun delete(keystore: String, alias: String) = CipherLock.withLock {
+        val keyStore = KeyStore.getInstance(keystore)
+        keyStore.load(null)
+        if (keyStore.containsAlias(alias)) keyStore.deleteEntry(alias)
+    }
+
+    private fun getKeyGenParameterSpecBuilder(alias: String): KeyGenParameterSpec.Builder = KeyGenParameterSpec
```

Comment:
Based on the above comments, you can move this to the body of `buildGCMKeySpec`

Reply:
3bc1ba8d443523dafcce732a1d79390218c1bb1d